contracts/V4Router.sol

```diff
@@ -0,0 +1,186 @@
+import {Currency, CurrencyLibrary} from "@uniswap/v4-core/src/types/Currency.sol";
+import {TickMath} from "@uniswap/v4-core/src/libraries/TickMath.sol";
+import {IHooks} from "@uniswap/v4-core/src/interfaces/IHooks.sol";
+import {PathKey} from "./libraries/PathKey.sol";
+import {IV4Router} from "./interfaces/IV4Router.sol";
+
+/// @title UniswapV4Router
+/// @notice Abstract contract that contains all internal logic needed for routing through Uniswap V4 pools
+abstract contract V4Router is IV4Router {
+    using CurrencyLibrary for Currency;
+
+    IPoolManager immutable poolManager;
+
+    constructor(IPoolManager _poolManager) {
+        poolManager = _poolManager;
+    }
+
+    function _v4Swap(SwapType swapType, bytes memory params) internal {
+        poolManager.unlock(abi.encode(SwapInfo(swapType, msg.sender, params)));
+    }
+
+    /// @inheritdoc IUnlockCallback
+    function unlockCallback(bytes calldata encodedSwapInfo) external override returns (bytes memory) {
+        if (msg.sender != address(poolManager)) revert NotPoolManager();
+
+        SwapInfo memory swapInfo = abi.decode(encodedSwapInfo, (SwapInfo));
+
+        if (swapInfo.swapType == SwapType.ExactInput) {
+            _swapExactInput(abi.decode(swapInfo.params, (IV4Router.ExactInputParams)), swapInfo.msgSender);
+        } else if (swapInfo.swapType == SwapType.ExactInputSingle) {
+            _swapExactInputSingle(abi.decode(swapInfo.params, (IV4Router.ExactInputSingleParams)), swapInfo.msgSender);
+        } else if (swapInfo.swapType == SwapType.ExactOutput) {
+            _swapExactOutput(abi.decode(swapInfo.params, (IV4Router.ExactOutputParams)), swapInfo.msgSender);
+        } else if (swapInfo.swapType == SwapType.ExactOutputSingle) {
+            _swapExactOutputSingle(abi.decode(swapInfo.params, (IV4Router.ExactOutputSingleParams)), swapInfo.msgSender);
+        } else {
+            revert InvalidSwapType();
+        }
+
+        return bytes("");
+    }
+
+    function _swapExactInputSingle(IV4Router.ExactInputSingleParams memory params, address msgSender) private {
+        _swap(
+            params.poolKey,
+            params.zeroForOne,
+            int256(-int128(params.amountIn)),
+            params.sqrtPriceLimitX96,
+            msgSender,
+            true,
+            true,
+            params.hookData
+        );
+    }
+
+    function _swapExactInput(IV4Router.ExactInputParams memory params, address msgSender) private {
+        unchecked {
+            uint256 pathLength = params.path.length;
+            uint128 amountOut;
+
+            for (uint256 i = 0; i < pathLength; i++) {
+                (PoolKey memory poolKey, bool zeroForOne) = _getPoolAndSwapDirection(params.path[i], params.currencyIn);
+                amountOut = uint128(
+                    _swap(
+                        poolKey,
+                        zeroForOne,
+                        int256(-int128(params.amountIn)),
```

Comment:
is this safe? it casts uint128 to int128 (unsafe), then flips, then goes to int256. Might be better to do
```
int256(-uint256(params.amountIn))
```
that feels safer?